sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -1766,19 +1766,20 @@ object CodeGenerator extends Logging {
       argSet += JavaCode.variable(ctx.INPUT_ROW, classOf[InternalRow])
     }
 
+    val stack = mutable.Stack[Expression](expr)
+
     // Collects local variables from a given `expr` tree
-    val collectLocalVariable = (ev: ExprValue) => ev match {
+    val collectLocalVariable = (ev: ExprValue, e: Expression) => ev match {
       case vv: VariableValue => argSet += vv
-      case _ =>
+      case _ => stack.pushAll(e.children)
     }
 
-    val stack = mutable.Stack[Expression](expr)
     while (stack.nonEmpty) {
       stack.pop() match {
         case e if subExprs.contains(e) =>
           val SubExprEliminationState(isNull, value) = subExprs(e)
-          collectLocalVariable(value)
-          collectLocalVariable(isNull)
+          collectLocalVariable(value, e)
+          collectLocalVariable(isNull, e)
```

Comment:
Oh, some `SubExprEliminationState` contains not `VariableValue` but still some statements that includes other expressions?

Reply:
> Oh, some `SubExprEliminationState` contains not `VariableValue` but still some statements that includes other expressions?

Yes, the expression contains input var  may match `SubExprEliminationState` and just skiped.

Reply:
Hmm, after second look, it looks not root cause. A subexpr should be evaluated before using it. 

So technically all its children expressions are not needed.

I found the root cause. `PromotePrecision` overwrites `Expression.genCode` where is subexpression replace happens. So `PromotePrecision` skips subexpression elimination. I will submit another PR and include your test there. Thanks.